ENH: Rolling rank

gsiano opened the pull request:
xref https://github.com/pandas-dev/pandas/issues/9481
- [x] closes #9481
- [x] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

gsiano committed:
ENH: rolling rank

rolling rank and percentile rank using skiplist

gsiano committed:
ENH: rolling rank

rolling rank and percentile rank using skiplist

gsiano committed:
Merge branch 'rolling_rank' of github.com:gsiano/pandas into rolling_rank

gsiano committed:
ENH: rolling rank

remove unused variables, fix indentation, add comment to roll_rank()

gsiano committed:
ENH: rolling rank

don't fill output when nobs < minp

pep8speaks commented:
Hello @gsiano! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/window/test_rolling.py`](https://github.com/pandas-dev/pandas/blob/4d06ba33e3f99876110312eb36912ca39e0ed910/pandas/tests/window/test_rolling.py):

> [Line 1504:1](https://github.com/pandas-dev/pandas/blob/4d06ba33e3f99876110312eb36912ca39e0ed910/pandas/tests/window/test_rolling.py#L1504): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1509:89](https://github.com/pandas-dev/pandas/blob/4d06ba33e3f99876110312eb36912ca39e0ed910/pandas/tests/window/test_rolling.py#L1509): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 88 characters)
> [Line 1514:1](https://github.com/pandas-dev/pandas/blob/4d06ba33e3f99876110312eb36912ca39e0ed910/pandas/tests/window/test_rolling.py#L1514): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1519:89](https://github.com/pandas-dev/pandas/blob/4d06ba33e3f99876110312eb36912ca39e0ed910/pandas/tests/window/test_rolling.py#L1519): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 88 characters)

gsiano committed:
ENH: rolling rank

address lint warnings

jreback commented on pandas/tests/window/test_rolling.py:
```diff
@@ -1500,3 +1500,31 @@ def test_rolling_numeric_dtypes():
+
+    result = Series([
+        ser[i - window:i].rank().iloc[-1] if i - window >= 0 else np.NaN
+        for i in range(1, length + 1)
```
can you compare vs .rolling.appyl(lambda x: x.rank()) vs this?

gsiano replied:
done

jreback reviewed (changes requested):
can you add this to the asvs as well.

jreback commented:
cc @mroeschke

jreback commented:
will this close https://github.com/pandas-dev/pandas/issues/9481 ? (can pull some examples fromt here)

jreback commented:
as a followup (can create an issue), need tests / impl for groupby.rolling.rank (it should just work though)

mzeitlin11 commented on pandas/_libs/window/aggregations.pyi:
```diff
@@ -63,6 +63,13 @@ def roll_quantile(
+    start: np.ndarray,
+    end: np.ndarray,
+    minp: int,
+    bint percentile,
```
```suggestion
    percentile: bool,
```

gsiano replied:
done

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1138,6 +1138,79 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+        int64_t nobs = 0, win
+        float64_t val
+        skiplist_t *skiplist
+        ndarray[float64_t] output
```
Since we define output from `np.empty`, this could have contiguity specified for small perf improvement, `float64_t[::1] output`

gsiano replied:
Does it make sense that I need to use `np.asarray` on `output` after making this change?

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1138,6 +1138,79 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+        raise MemoryError("skiplist_init failed")
+
+    with nogil:
+        for i in range(0, N):
```
```suggestion
        for i in range(N):
```

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1138,6 +1138,79 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+            if i == 0 or not is_monotonic_increasing_bounds:
+                if not is_monotonic_increasing_bounds:
+                    nobs = 0
+                    skiplist = skiplist_init(<int>win)
```
If we just reassign the pointer here, is it possible that we leak the previously pointed to skiplist?

gsiano replied:
yeah this is a leak. I added a destroy call to get rid of it. This is also an issue in some of the other rolling functions. Should I fix those in this pr or make a separate issue for it?

mzeitlin11 replied:
Best to make a separate issue since this will already be a pretty large pr (and smaller the pr, the easier to review)

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1138,6 +1138,79 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+                        rank = skiplist_insert(skiplist, val)
+            if nobs >= minp:
+                if percentile:
+                    output[i] = <float64_t>(rank + 1) / nobs if rank != -1 else NaN
```
Why do we return `NaN` if `rank == -1`? Wouldn't that only happen if `skiplist_insert` failed because of a memory error?

gsiano replied:
I changed this to raise MemoryError when the insert fails

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1138,6 +1138,79 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+                if percentile:
+                    output[i] = <float64_t>(rank + 1) / nobs if rank != -1 else NaN
+                else:
+                    output[i] = rank + 1 if rank != -1 else NaN
```
Same comment as above

gsiano replied:
changed this to raise MemoryError when the insert fails

mzeitlin11 reviewed (commented):
Thanks for the pr @gsiano! Some comments on the cython code

gsiano committed:
ENH: rolling rank

raise MemoryError on skiplist_insert failure, destroy skiplist before
re-init, and other minor changes

mroeschke commented on pandas/core/window/rolling.py:
```diff
@@ -1409,6 +1409,14 @@ def quantile(self, quantile: float, interpolation: str = "linear", **kwargs):
 
         return self._apply(window_func, name="quantile", **kwargs)
 
+    def rank(self, pct: bool = False, **kwargs):
```
Bonus points if you could implement `ascending` to match `DataFrame.rank`

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rank.html#pandas.DataFrame.rank

mroeschke commented:
What happens currently when there are ties?

`DataFrame.rank` currently has a `method` argument to control that behavior https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rank.html#pandas.DataFrame.rank

gsiano committed:
ENH: rolling rank - rank methods

implement min, max, and average rank methods

gsiano commented:
> What happens currently when there are ties?
> 
> `DataFrame.rank` currently has a `method` argument to control that behavior https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rank.html#pandas.DataFrame.rank

I added the min, max, and average rank methods. I still have to add the dense and first methods, and I'll look into adding the other flags from DataFrame.rank

gsiano commented:
> I added the `min`, `max`, and `average` rank methods. I still have to add the `dense` and `first` methods, and I'll look into adding the other flags from `DataFrame.rank`

the `first` method doesn't seem relevant for rolling rank, and `dense` probably deserves a separate pr

mzeitlin11 commented:
> > I added the `min`, `max`, and `average` rank methods. I still have to add the `dense` and `first` methods, and I'll look into adding the other flags from `DataFrame.rank`
> 
> the `first` method doesn't seem relevant for rolling rank, and `dense` probably deserves a separate pr

+1 on pushing `dense` to a separate pr. `ascending` would be a nice flag to have ... I wonder if something like just adding the negative of the value to the `skiplist` would handle flipping the order

gsiano commented:
> +1 on pushing `dense` to a separate pr. `ascending` would be a nice flag to have ... I wonder if something like just adding the negative of the value to the `skiplist` would handle flipping the order

That sounds like it will work. I'll try adding it.

gsiano committed:
ENH: rolling rank - `ascending` flag

added the `ascending` flag, various cleanups, expanded tests and asv benchmark

gsiano committed:
ENH: rolling rank

remove unimplemented rank types

mroeschke commented on pandas/core/window/rolling.py:
```diff
@@ -1409,6 +1409,16 @@ def quantile(self, quantile: float, interpolation: str = "linear", **kwargs):
 
         return self._apply(window_func, name="quantile", **kwargs)
 
+    def rank(self, pct: bool = False, method: str = "average", ascending: bool = True, **kwargs):
```
Nit: Could you have this order mirror `DataFrame.rank` as close as possible i.e. `rank(self, method, ascending, pct)`

Also is `na_option` applicable here like in `DataFrame.rank`: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rank.html

mroeschke replied:
I see I think `na_options` may be hard to resolve with `min_periods` since both key words are supposed to handle NaNs, so maybe it's okay to leave out `na_options` for now.

The `rolling.apply` behavior is expected due to the `min_periods` behavior

```
min_periods : int, default None

    Minimum number of observations in window required to have a value (otherwise result is NA). For a window that is specified by an offset, min_periods will default to 1. Otherwise, min_periods will default to the size of the window.

```

gsiano committed:
ENH: rolling rank - reorder parameter list

reorder parameter list to match DataFrame.rank

gsiano committed:
ENH: rolling rank - address pre-commit errors

jreback commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1139,6 +1141,116 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+                            raise MemoryError("skiplist_insert failed")
+                        if rank_type == AVERAGE:
+                            rank_min = skiplist_min_rank(skiplist, val)
+                            rank = ((rank * (rank + 1) / 2)
```
can you use parens rather than line continuation? we typically do this in most places

jreback reviewed (changes requested):
looks good, can you add
- whatsnew note (1.4 enhacements), wouldn't object to a small example if you'd like (or a single note ok)
- add in the reference section: https://github.com/pandas-dev/pandas/blob/master/doc/source/reference/window.rst somewhere
- this is *just* for rolling, or works for expanding? (i think it should, can you add tests)
- i think we need a doc-string in rolling.py

gsiano committed:
ENH: rolling rank

added tests for `Expanding`. added doc strings and whatsnew note

gsiano commented:
> looks good, can you add
> 
> * whatsnew note (1.4 enhacements), wouldn't object to a small example if you'd like (or a single note ok)
> * add in the reference section: https://github.com/pandas-dev/pandas/blob/master/doc/source/reference/window.rst somewhere
> * this is _just_ for rolling, or works for expanding? (i think it should, can you add tests)
> * i think we need a doc-string in rolling.py

I added a test for expanding and updated the docs and whatsnew file. Is there a way to render the docs locally so I can see how it looks?

gsiano committed:
ENH: rolling rank - fix pre-commit errors

mroeschke commented:
> Is there a way to render the docs locally so I can see how it looks?

https://pandas.pydata.org/pandas-docs/stable/development/contributing_documentation.html#how-to-build-the-pandas-documentation

mroeschke commented on pandas/tests/window/test_expanding.py:
```diff
@@ -264,3 +264,25 @@ def test_expanding_skew_kurt_numerical_stability(method):
+@pytest.mark.parametrize("ascending", [True, False])
+@pytest.mark.parametrize("test_data", ["default", "duplicates", "nans"])
+def test_rank(window, method, pct, ascending, test_data):
+    length = 1000
```
Nit: Could we test with a smaller sample (e.g. 20) and adjust `window` accordingly?

mroeschke commented on pandas/tests/window/test_rolling.py:
```diff
@@ -1500,3 +1500,25 @@ def test_rolling_numeric_dtypes():
+@pytest.mark.parametrize("ascending", [True, False])
+@pytest.mark.parametrize("test_data", ["default", "duplicates", "nans"])
+def test_rank(window, method, pct, ascending, test_data):
+    length = 1000
```
Same

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1139,6 +1141,120 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+                    else:
+                        rank = NaN
+            if nobs >= minp:
+                output[i] = <float64_t>(rank) / nobs if percentile else rank
```
Is the cast here necessary?

mzeitlin11 commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1139,6 +1141,120 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
+                            raise MemoryError("skiplist_insert failed")
+                        if rank_type == AVERAGE:
+                            rank_min = skiplist_min_rank(skiplist, val)
+                            rank = (((rank * (rank + 1) / 2)
```
Would you mind adding a small comment explaining the logic of this formula?

mzeitlin11 commented on pandas/core/window/expanding.py:
```diff
@@ -564,6 +564,81 @@ def quantile(
+    )
+    def rank(
+        self,
+        method: str = "average",
```
To be consistent this could be the same `Literal` you defined in `aggregations.pyi`? Maybe that could be aliased in `pandas/_typing.py`, so that then if we do something like add `dense` then we only need to change typing in one place.

jreback replied:
yeah these should be consisteent across the rank methods (could be a followup to fix this)

mzeitlin11 commented on pandas/tests/window/test_expanding.py:
```diff
@@ -264,3 +264,25 @@ def test_expanding_skew_kurt_numerical_stability(method):
+    elif test_data == "duplicates":
+        ser = Series(data=np.random.choice(3, length))
+    elif test_data == "nans":
+        ser = Series(data=np.random.choice([1.0, 0.25, 0.75, np.nan], length))
```
Can you throw an `inf` and `-inf` in here? Not really a special case I guess with how it's implemented in `rolling_rank`, but it is special cased in other rank algos and were some issues before with `inf` and `nan`, so covering here would be good

mzeitlin11 commented on pandas/tests/window/test_rolling.py:
```diff
@@ -1500,3 +1500,25 @@ def test_rolling_numeric_dtypes():
+    elif test_data == "duplicates":
+        ser = Series(data=np.random.choice(3, length))
+    elif test_data == "nans":
+        ser = Series(data=np.random.choice([1.0, 0.25, 0.75, np.nan], length))
```
Same comment as above about `inf`

mzeitlin11 reviewed (approved):
Small comments, otherwise LGTM!

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -87,6 +87,53 @@ Multithreaded CSV reading with a new CSV Engine based on pyarrow
+
+.. ipython:: python
+
+    >>> s = pd.Series([1, 4, 2, 3, 5, 3])
```
just put the commands; they will render during the doc-build (e.g. L100 w/o the '>>>' and not below)

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -87,6 +87,53 @@ Multithreaded CSV reading with a new CSV Engine based on pyarrow
+    5    2.0
+    dtype: float64
+
+    >>> s.expanding().rank()
```
you could add a comment; this works for expanding as well (or kill the expanding, up 2 you)

jreback commented on pandas/_libs/src/skiplist.h:
```diff
@@ -180,10 +180,28 @@ PANDAS_INLINE double skiplist_get(skiplist_t *skp, int i, int *ret) {
     return node->value;
 }
 
+PANDAS_INLINE int skiplist_min_rank(skiplist_t *skp, double value) {
```
worth a comment here on what this does?

jreback commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1139,6 +1141,120 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
     return output
 
 
+cdef enum RankType:
```
can you use the enums defined in pandas/_libs/algos.pyx? e.g. the TIEBREAK_AVERAGE. may need to move the to algox.pyd to import properly.

jreback commented on pandas/core/window/rolling.py:
```diff
@@ -1409,6 +1409,22 @@ def quantile(self, quantile: float, interpolation: str = "linear", **kwargs):
 
+    def rank(
+        self,
+        method: str = "average",
```
same comment here

gsiano committed:
ENH: rolling rank

addressing comments

gsiano committed:
Merge branch 'master' into rolling_rank

gsiano committed:
ENH: rolling rank

add rolling rank tiebreakers dict

gsiano committed:
ENH: rolling rank

fix docs warnings - remove '>>>' from code block

jreback commented on pandas/_libs/window/aggregations.pyx:
```diff
@@ -1139,6 +1143,122 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
     return output
 
 
+rolling_rank_tiebreakers = {
```
possible to unify these with the same in algos.pyx?

jreback reviewed (approved):
lgtm. small comment to consider for followup.

cc @mzeitlin11 over to you

jreback commented:
oh see you approved, ok then.

jreback commented:
thanks @gsiano very nice! pls create an issue for for the requested followups (PR as well if you can .....) keep em coming!
